src/coreclr/jit/morph.cpp

```diff
@@ -10869,7 +10869,71 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
             INDEBUG(node->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
             return node;
         }
+#if defined(TARGET_XARCH)
+        case NI_SSE_And:
+        case NI_SSE2_And:
+        case NI_AVX_And:
+        case NI_AVX2_And:
+        {
+            if (node->GetOperandCount() != 2)
+            {
+                return node;
+            }
+
+            GenTree*            op1      = node->Op(1);
+            GenTree*            op2      = node->Op(2);
+            GenTree*            lhs      = nullptr;
+            GenTree*            rhs      = nullptr;
+            GenTreeHWIntrinsic* inner_hw = nullptr;
+
+            // Transforms ~v1 & v2 to VectorXxx.AndNot(v2, v1)
+            if (op1->OperIs(GT_HWINTRINSIC))
+            {
+                rhs      = op2;
+                inner_hw = op1->AsHWIntrinsic();
+            }
+            // Transforms v2 & (~v1) to VectorXxx.AndNot(v1, v2)
+            else if (op2->OperIs(GT_HWINTRINSIC))
+            {
+                rhs      = op1;
+                inner_hw = op2->AsHWIntrinsic();
+            }
+            else
+            {
+                return node;
+            }
```

Comment:
This is going to miss the optimization for cases like: `((x & y) & ~z)`

You're going to need to check that it is a hwintrinsic and that it is the relevant `xor` (xarch and arm) or `not` (arm only) node.

Reply:
There is also potentially a concern around `side effects` and ensuring that `~x & y`, which must be represented as `gtNewSimdBinOpNode(AND_NOT, y, x, ...)` preserves side effects with regards to `x` being evaluted before `y`.

Reply:
I have resolved some comments and pushed them to make sure I got you right.
Could you please give me a hint how to treat `not` for arm and how to test it and how to handle the sideeffect case?